server/src/main/java/org/opensearch/cluster/ApplicationManager.java

```diff
@@ -0,0 +1,153 @@
+        }
+
+        return Set.of();
+    }
+
+    /**
+     * Checks scopes of an application subject and determine if it is allowed to perform an operation based on the given scopes
+     * @param wrapped The ApplicationSubject whose scopes should be evaluated
+     * @param scopes The scopes to check against the subject
+     * @return true if allowed, false if none of the scopes are allowed.
+     */
+    public boolean isAllowed(ApplicationSubject wrapped, final List<Scope> scopes) {
+
+        final Optional<Principal> optionalPrincipal = wrapped.getApplication();
+
+        if (optionalPrincipal.isEmpty()) {
+            // If there is no application, actions are allowed by default
+            return true;
+        }
+
+        if (!wrapped.applicationExists()) {
+            return false;
+        }
+
+        final Set<String> scopesOfApplication = wrapped.getScopes();
+
+        boolean isApplicationSuperUser = scopesOfApplication.stream()
+            .map(Scope::parseScopeFromString)
+            .anyMatch(parsedScope -> parsedScope.equals(ApplicationScope.SUPER_USER_ACCESS));
+
+        if (isApplicationSuperUser) {
+
+            return true;
+        }
+
+        boolean isMatchingScopePresent = scopesOfApplication.stream()
+            .map(Scope::parseScopeFromString)
+            .anyMatch(parsedScope -> scopes.stream().anyMatch(parsedScope::equals));
+
+        return isMatchingScopePresent;
+    }
+
+    /**
+     * This method allows for checking the ExtensionPointScopes associated with an application
+     * @param principal The principal of the application whose scopes you want to check
+     * @return A set of all extension point scopes associated with that principal
+     */
+    public Set<String> getExtensionPointScopes(Principal principal) {
+
+        if (extensionManager.getExtensionPrincipals().contains(principal)) {
+            return extensionManager.getExtensionIdMap()
+                .get(principal.getName())
+                .getScopes()
+                .stream()
+                .filter(scope -> Scope.parseScopeFromString(scope).getNamespace() == ScopeEnums.ScopeNamespace.EXTENSION_POINT)
+                .collect(Collectors.toSet());
+        }
+
+        return Set.of();
+    }
+
+    /**
+     * Checks whether there is an application associated with the given principal or not
+     * @param principal The principal for the application you are trying to find
+     * @return Whether the application exists (TRUE) or not (FALSE)
+     */
+    public boolean associatedApplicationExists(Principal principal) {
+        return (extensionManager.getExtensionPrincipals().contains(principal));
+    }
+
+    /**
+     * Allows for checking the ExtensionManager being used by the ApplicationManager in the case that there are multiple application providers
+     * @return The ExtensionManager being queried by the ApplicationManager
+     */
+    public ExtensionsManager getExtensionManager() {
```

Comment:
Getters/Setters should be used on data objects, this is a business logic class, it shouldn't have any exposed fields, its purpose is to isolate its specific understanding of Principal -> 'Application' -> Scopes that application has

Reply:
This ApplicationManager setup is being used to funnel the different types of Applications together. Specifically, it will wrap the ExtensionManager and a future PluginManager to allow handling of both types of applications with the same manager. All of the getters being used are based on the wrapped ExtensionManager (and in the future a wrapped PluginManager). I do not know what would be an appropriate alternative to this setup if we want to make a "funnel" which can manage both extensions and plugins. We want to be able to have a single place where all different concepts of Applications can be operated on as the same regardless of their minor differences.